Add viewer tool

s9latimm opened the pull request:
M	CMakeLists.txt
M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Add viewer tool

M	CMakeLists.txt
M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Remove ports

M	.gitignore
M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Improve graph writer

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
No border, fixed ordering

M	include/vara/Feature/FeatureModel.h

vulder commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -191,6 +175,26 @@ struct DOTGraphTraits<vara::feature::FeatureModel *>
   }
 
+  static void buildCluster(llvm::raw_ostream &O, vara::feature::Feature *Node,
+                           int L = 1) {
```
what is `L` could you please add a comment for the function that also describes the parameters

vulder commented on tools/fm-viewer/FeatureModelViewer.cpp:
```diff
@@ -1,20 +1,81 @@
-      llvm::ViewGraph(FM.get(), llvm::Twine(FM->getName()));
-      return 0;
+
+      if (FM) {
+        if (Dump) {
+          FM->dump();
+        } else {
+          std::string Filename =
+              llvm::WriteGraph(FM.get(), llvm::Twine(FM->getName()));
+          if (!Filename.empty()) {
+            if (llvm::ErrorOr<std::string> P =
+                    Viewer.empty() ? llvm::errc::invalid_argument
+                                   : llvm::sys::findProgramByName(Viewer)) {
+              llvm::errs() << "Trying '" << *P << "' program... \n";
+              llvm::sys::ExecuteAndWait(*P, {*P, Filename});
+            } else {
+              llvm::DisplayGraph(Filename);
+            }
+            llvm::errs() << "Remember to erase graph file: " << Filename
+                         << "\n";
+          }
+        }
+        return 0;
+      }
+    } else {
+      llvm::errs() << "error: Could not open file.\n";
     }
+  } else {
+    llvm::errs() << "error: Expected single file.\n";
```
We should restructure this to make error checking and return and not nest everything because this becomes hard to reason about. For example:
```
if (!FM) {
  llvm::errs() << "Foo not feature model\n";
  return 1;
}
... continue the good case
```
Otherwise, we create on nesting depth per error case.

vulder commented on tools/fm-viewer/FeatureModelViewer.cpp:
```diff
@@ -1,20 +1,81 @@
+  llvm::InitLLVM X(Argc, Argv);
+
+  llvm::cl::HideUnrelatedOptions(FMViewerCategory);
+
```
```suggestion
```

vulder commented on tools/fm-viewer/FeatureModelViewer.cpp:
```diff
@@ -1,20 +1,81 @@
+
+  llvm::cl::HideUnrelatedOptions(FMViewerCategory);
+
+  llvm::cl::ParseCommandLineOptions(Argc, Argv,
+                                    "Shows feature model as graph.");
```
```suggestion
  llvm::cl::ParseCommandLineOptions(Argc, Argv,
                                    Overview, /*Errs=*/nullptr, FlagsEnvVar);
```

vulder commented on tools/fm-viewer/FeatureModelViewer.cpp:
```diff
@@ -1,20 +1,81 @@
+
+int main(int Argc, char **Argv) {
+  llvm::InitLLVM X(Argc, Argv);
+
```
```suggestion
  const char *FlagsEnvVar = "FM_VIEWER_FLAGS";
  const char *Overview = R"(Please add overview/intro text here)"
```

vulder commented on tools/fm-viewer/FeatureModelViewer.cpp:
```diff
@@ -1,20 +1,81 @@
+      if (Xml) {
+        std::string DocRaw((std::istreambuf_iterator<char>(DocFile)),
+                           std::istreambuf_iterator<char>());
+        FM = vara::feature::FeatureModelXmlParser(DocRaw).buildFeatureModel();
```
```suggestion
        FM = vara::feature::FeatureModelXmlParser(std::move(DocRaw)).buildFeatureModel();
```

vulder commented on tools/fm-viewer/FeatureModelViewer.cpp:
```diff
@@ -1,20 +1,81 @@
+        } else {
+          std::string Filename =
+              llvm::WriteGraph(FM.get(), llvm::Twine(FM->getName()));
+          if (!Filename.empty()) {
```
Is empty a usual thing that can happen here? If not I would suggest adding and assert and `return 1` with an error message;

s9latimm replied:
Returning without error mimics the behavior of `llvm::ViewGraph` (`GraphWriter.h`) given `WriteGraph` failed to open a file for writing (which may be kind of usual). So instead, I would terminate with an error - as the tool basically failed.

s9latimm committed:
Use html label

M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/Location.h

s9latimm committed:
Requested changes

M	include/vara/Feature/FeatureModel.h
M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Fix repeated parsing

M	include/vara/Feature/FeatureModelParser.h
M	lib/Feature/FeatureModelParser.cpp

s9latimm committed:
Include MemoryBuffer

M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Write own graph writer

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Requested changes

M	include/vara/Feature/Feature.h
M	include/vara/Feature/Relationship.h
M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Use NoneType

M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Merge branch 'f-FMViewer' into f-GraphWriter

M	include/vara/Feature/Feature.h
M	include/vara/Feature/Relationship.h
M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Working prototype

M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h

s9latimm committed:
WS

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Fix StringRef

M	bindings/python/tests/feature/test_feature.py
M	bindings/python/tests/feature_model/test_feature_model.py

s9latimm committed:
Add comments

M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Merge branch 'vara-dev' into f-FMViewer

M	bindings/python/tests/TEST_INPUTS/example_feature_model.xml
M	bindings/python/tests/feature/test_feature.py
A	bindings/python/tests/feature/test_location.py
M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/vara-feature/CMakeLists.txt
M	bindings/python/vara-feature/pybind_Feature.cpp
M	bindings/python/vara-feature/pybind_FeatureModel.cpp
M	bindings/python/vara-feature/pybind_Init.cpp
A	bindings/python/vara-feature/pybind_Location.cpp
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelParser.h
M	include/vara/Feature/Location.h
M	lib/Feature/FeatureModelParser.cpp
M	unittests/Feature/BinaryFeature.cpp
M	unittests/Feature/CMakeLists.txt
A	unittests/Feature/Feature.cpp
M	unittests/Feature/Location.cpp

s9latimm committed:
Fix iterators

M	bindings/python/vara-feature/pybind_Feature.cpp
M	include/vara/Feature/Feature.h
M	unittests/Feature/BinaryFeature.cpp

s9latimm committed:
Add target

M	.travis.yml
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/Location.h
M	tools/fm-viewer/CMakeLists.txt

vulder commented on .travis.yml:
```diff
@@ -4,7 +4,7 @@ os: linux
 
 before_install:
-  - sudo apt-get -y install libstdc++6 libxml2-dev
+  - sudo apt-get -y install libstdc++6 libxml2-dev graphviz
```
we should add this dep also to the readme

vulder commented on .travis.yml:
```diff
@@ -67,7 +67,7 @@ jobs:
             - clang-tidy-10
       env: CXX=g++-9 BUILD_TYPE=Debug
-    - name: GCC-9 Release
+    - name: GCC-9 Release\
```
What is the meaning of this `\`?

vulder committed:
Merge branch 'vara-dev' into f-FMViewer

s9latimm committed:
Add check and verify

M	.travis.yml
M	README.md
M	lib/Feature/FeatureModelParser.cpp
M	tools/fm-viewer/CMakeLists.txt
M	tools/fm-viewer/FeatureModelViewer.cpp

s9latimm committed:
Emit comments

M	include/vara/Feature/FeatureModel.h

vulder committed:
trigger travis
